### ISSUE TITLE
Rules E712,F632 causes infinite loop

### ISSUE THREAD

**Issue by qarmin** [astral-sh/ruff#6891]
Ruff 0.0.286 (latest changes from main branch)

```
ruff  *.py --select E712,F632 --no-cache
```

file content:
```
class RefsContainerTests:
        f.write(
            b"\n".join(
                line
                for line in refs_data.split(b"\n")
                if not line or line[0] not in b"#^" is not True       ),
        )
```

error:
```
error: Failed to converge after 100 iterations.

This indicates a bug in `ruff`. If you could open an issue at:

    https://github.com/astral-sh/ruff/issues/new?title=%5BInfinite%20loop%5D

...quoting the contents of `PY_FILE_TEST_1518432890.py`, the rule codes D100, D101, F632, F821, PLR0133, along with the `pyproject.toml` settings and executed command, we'd be very appreciative!
```

[PY_FILE_TEST_1518432890.py.zip](https://github.com/astral-sh/ruff/files/12444780/PY_FILE_TEST_1518432890.py.zip)

**Comment by dhruvmanila** [comment]
Thanks! I've narrowed it down to:

```python
if foo not in "foo" is not True:
    pass
```

with the command:

```
cargo run --bin ruff -- check --no-cache --isolated --select=E712,F632 t.py --fix
```

**Comment by qarmin** [comment]
File no longer cause problems
